### ISSUE TITLE
Documentation support: New documentation for incremental & staging launches

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-team#93032]
### Request type

- [X] Writing new documentation
- [ ] Updating previously published documentation
- [ ] Reviewing recently written, never-before-published drafts

### Number of docs

1

### Subject matter expert

Jill Adams

### Embedded media

- [ ] Screenshot(s)
- [ ] Diagram(s)
- [ ] Video(s)

### Audience

- [X] Developers
- [ ] Designers and researchers
- [ ] Product managers
- [ ] Other

### Anything else?

Context for why this content belongs in the DEPO site: https://dsva.slack.com/archives/C03LFSPGV16/p1726518596517349?thread_ts=1726518589.434909&cid=C03LFSPGV16

Reference document that contains content that should move into the Content site: https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/public-websites#shadow--dark-launches-of-content.  

This info is primarily about:
1. Shipping new features using % rollout to unauthenticated users, since the standard Flipper implementation can only be used for authed users
2. Shipping CMS-driven content to Staging only for pre-production preview and usability testing

The github doc references the matrix of what is and isn't possible, for staging features pre-production, depending on the tech stack. I think that info is also valuable to maintain for awareness in the broader environment.

**Comment by jilladams** [comment]
Since this info is nested in a giant github doc about many other things, here's the section we want to figure out how to migrate for broader awareness: 

## Shadow / dark launches of content
Docs below will move to Platform content site via https://github.com/department-of-veterans-affairs/va.gov-team/issues/93032. When that is published, content will be removed from this README.

### Ownership: CMS team
**CMS Dark launches are owned by the CMS team as of 10/31/2023.**  

### Usage
Sometimes content generated by the Sitewide Content team, other teams, or just portions of a page may need to “dark launch” or incrementally launch, meaning: not be available to all site users in production. We have options for how to manage dark launches: 

### It is **not** currently possible to
* Hide a specific CMS node from robots.txt & the sitemap. Not ticketed, and would/will require custom code.
* Display a CMS-generated page to a % of users for incremental launch.

### It IS currently possible to

#### 1. Use Tugboat to preview CMS + content-build + vets-website changes together.
* In the `/admin/content/deploy` Release Content interface on Tugboat, you can specify a content-build and/or a vets-website branch, to get a full feature preview in Tugboat frontend.

**Not working as of May 2024**: Tugboat is not picking up vets-website changes from branches other than `main`. Ticketed pending CMS ability to prioritize a fix: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/18208

#### 2. Use vets-api “Flipper” feature toggles to show / hide page sections or react widgets (vets-website changes) to authenticated users

- https://api.va.gov/flipper/features – to login: you must have a verified ID.me account using your VA.gov email address (as primary or secondary email) 
- https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles 
- https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide 

Flippers cannot control visibility of an entire page to unauthenticated users or prevent it from being accessed. Once a page exists in production, Flipper can control visibility of the element it's wrapped around. Flipper controls can show/hide:

- boolean style (on/off) for the wrapped element
- visibility to % of logged in users. Used for slow rollouts e.g. to 25%, 50%, 100% of page visitors
- visibility % time, which we believe means page serves, e.g. 25% of times the page is loaded, whatever is behind the flipper will be displayed. 
- Logged-in Actors, specified by email address, who will / won't see the content (can also specify a % of Actors)

A good summary of how to flipper: https://github.com/department-of-veterans-affairs/va.gov-team/issues/8853#issuecomment-625426153 

Flippers are typically using javascript, and do not require a code deploy to enable / change settings.

#### 3. Write custom vets-website code to control traffic to unauthenticated element / widget
A react hook exists that can allow displaying a feature to a % of unauthenticated traffic:  `useStaggeredFeatureRelease`
https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/utilities/react-hooks.js#L34

This does not work for CMS-generated content. 

#### 4. Publish a CMS page only to Staging (using EntityQueue) in order to stage CMS + content-build + vets-website for viewing off of VA network
Typically, when a CMS page is published, it publishes to production. 
However: as of Sept 2022, a CMS node may be published to staging.va.gov _only_ by using EntityQueue. This is useful to allow previewing a new CMS page that includes a React widget (aka vets-website code) for user acceptance testing (by off-network users) or stakeholder approval, before it reaches production. CMS team does not want to get in the habit of having staged nodes that will not publish, so this method should only be used for business cases that warrant it, at CMS PO's discretion.

**How to stage:** 
* (Requesting team) cut a [Dark launch request ticket](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/new/choose) to CMS team to request support
  * Specify Node / prod CMS link to the node 
* (Requesting team) When CMS has prioritized the work, merge React widget code to main & confirm deploy in vets-website. App code will not appear on prod, **if no live page is calling the app**.
* (CMS) Add the Node ID to the [Staged Content entity subqueue](https://prod.cms.va.gov/admin/structure/entityqueue/staged_content/staged_content?destination=/admin/structure/entityqueue)
* (CMS) Request approval from requesting team to publish the node to staging, then publish it.
* (CMS/Deployment) 
  * The nightly CMS build must run which pushes prod mirror to the staging site, or CMS must request an out of band deployment. 
  * Content-release must then run. Page will be live on Staging after the first content-release after nightly CMS build.

**How to launch**
* (Requesting team) Greenlight to launch
* (CMS) Remove the node ID from the [Staged Content entity subqueue](https://prod.cms.va.gov/admin/structure/entityqueue/staged_content/staged_content?destination=/admin/structure/entityqueue)
* (CMS/Deployment) 
  * The nightly CMS build must run or CMS must run an out of band deployment. 
  * Content-release must then run. Page will be live on Prod after the first content-release after nightly CMS build.

NOTES: 
1. It is _not_ possible to use this mechanism to test updates / additions to an existing live/published CMS page. The dark launch mechanism says: for this node ID, include it in the Staging build, and do _not_ include it in the Production build. So testing an existing node in this way would effectively unpublish it from Production.
2. For reference, the staging CMS database is not implicated in this process / these changes. The front-end of staging.va.gov is populated by a database mirror of Prod, and is rewritten everytime code is merged to prod. The staging CMS, staging.cms.va.gov, has no front-end -- no front-end is updated when the Staging CMS db is modified.

**Entityqueue implementation:**
- Spike: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10327
- FE: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10374
- BE: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10375


#### 5. Use CMS feature flags to show / hide fields or portions of templates in content-build
- CMS docs https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/READMES/interfaces.md#featureflags 
- Front-end usage docs: https://github.com/department-of-veterans-affairs/vspwiki/blob/master/frontend/pages/platform/tools/feature-toggles.md#testing-cms-feature-toggles 

CMS Feature flags can control visibility from the template layer, but cannot control query content / behavior.
CMS feature toggles take effect at the content-build level, and are not dynamically controlled with javascript, so do require a content release to take effect in prod.

For example, if a new feature is launching, that calls new fields / exposes new layout of a template, we may use CMS feature flags to wrap the new markup in order to show / hide the changes for testing or to simplify rollback. 

We try to avoid page-specific templates, so for example, we would not advocate controlling launch of content for a single page using this method. This is a content-type-level option.


#### 6. Use content-build environment checks to publish a static page only to staging
This only applies to static pages, not CMS content. In content-build, we can exclude a static page from production builds.

PW team has done this twice: 
1. Published a static page for Harassment Reporting Tool, [#35161 [FE] Build static landing page in Staging environment](https://github.com/department-of-veterans-affairs/va.gov-team/issues/35161)
2. Published a homepage prototype for UA testing, /homepage-test: [#8340 Iterate VA.gov homepage changes for demo/testing in publicly accessible environment](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8340); [PR](https://github.com/department-of-veterans-affairs/content-build/pull/1081/files#diff-9d481a9d1d5399befb7b3d6424a262b6e38796b248c017c0f6bacddd4bd54814) 

**Gotcha:** this method creates tech debt, and can cause unexpected behavior in future, when we test in Staging and end up with different behavior in Production. E.g. Feature A tests well in staging, but when we reach Prod, behavior is different due to a forgotten environment check in old / untouched code.


Some nuances: https://dsva.slack.com/archives/CT4GZBM8F/p1660937436605269

**Comment by jknipes** [comment]
Hi @jilladams thanks for your content suggestions. The Platform Content team will review your request and put it in our backlog to refine for next steps.

**Comment by jilladams** [comment]
Randomly was thinking about IA today: I think there's a bigger picture thing in the ecosystem around: 
1. How can I show actual app code to users for testing, for features that aren't vets-api > vets-website driven, where feature toggles won't work? and 
2. How can I incrementally ship a feature, same scenario? (Where feature toggles aren't an option.)

All the topics here circle around that. So my thinking so far is something like: 

![Screenshot 2024-09-18 at 3 14 36 PM](https://github.com/user-attachments/assets/a071b40e-92fb-4a13-973b-9a9bc4baf30e)

Do with that as you please. ;)

**Comment by jknipes** [comment]
Hi @jilladams , thanks for your content suggestions. The Platform Content team will review your request and put it in our backlog to refine for next steps.

**Comment by jknipes** [comment]
@gracekretschmer-metrostar and @jilladams we were unable to plan this into Sprint 11 due to 

1. prioritizing of Redirect Implementation Strategy page ([93073](https://github.com/department-of-veterans-affairs/va.gov-team/issues/93073), [93557](https://github.com/department-of-veterans-affairs/va.gov-team/issues/93557), [93560](https://github.com/department-of-veterans-affairs/va.gov-team/issues/93560))
2. previously planned work with Product team ([92335](https://github.com/department-of-veterans-affairs/va.gov-team/issues/92335), [92345](https://github.com/department-of-veterans-affairs/va.gov-team/issues/92345))
3. unexpected early offboarding of core team member

**Comment by jknipes** [comment]
cc: @edmund-dunn

**Comment by jknipes** [comment]
(Previous comment accidently deleted) We were unable to begin this project due to prioritization of migrating [the Redirect Implementation Strategy page ](https://github.com/department-of-veterans-affairs/va.gov-team/issues/93073)with Sitewide Websites, and [the Partner Services project](https://github.com/department-of-veterans-affairs/va.gov-team/issues/91769) with Platform Product team. 
We are proceeding with a review of the Incremental & Staging Launches content from Edmund Dunn, per [this Slack thread](https://dsva.slack.com/archives/C03LFSPGV16/p1726574630035839?thread_ts=1726518589.434909&cid=C03LFSPGV16), and we are hoping to plan the edits in Sprint 12.

**Comment by jilladams** [comment]
Noting I learned something today: 

#2 above says Flippers can only be used for authed users
#3 Says you can do stuff for unauthed with a vets-website react hook

Except it seems that apparently you can also use Flippers to show FE-only (not backend) changes to unauthed users, if you use the `cookie_id` and % of time setting in Flipper. The docs on the Feature Toggles page sort of address this but the language is sort of hard to parse, to know that's true with confidence: 

https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-3.Determinehowyouwantthefeaturetoggletobe%22sticky%22.1
> For the behavior to be consistent for a user for the duration of a cookie within a single browser, regardless of their logged in status choose cookie_id as the actor_type.

https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-2.ConfigureFlipperUIsettings
> If the feature toggle's actor_type in [config/features.yml](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/features.yml) is set to cookie_id rather than user, the feature instead applies to only a percentage of users for the duration of a cookie within a single browser, regardless of their logged in status.
Setting actor_type to cookie_id ONLY AFFECTS THE FRONTEND, NOT THE BACKEND.
> This is useful when you need to apply a staged rollout that involves an unauthenticated user experience.

> **Not Logged-in Users**
> If the feature toggle’s actor_type in [config/features.yml](https://github.com/department-of-veterans-affairs/vets-api/blob/master/config/features.yml) is set to cookie_id rather than user, the feature instead applies to only a percentage of users for the duration of a cookie within a single browser, regardless of their logged in status. This is useful when you need to apply a staged rollout that involves an unauthenticated user experience.

> Percentage of Time | Enable a feature for all users for a percentage of time.
 

I think that means I should not be considered the SME here for anything other than the CMS dark launch capabilities (😂 ) and a FE engineer deeply familiar with Flipper functionality should be a technical reviewer of any content changes related to Flipper.

**Comment by jknipes** [comment]
Thanks @jilladams , we should create a separate ticket to update and clarify the Flipper functionality content, and this ticket will be confined to the Dark Launch content

**Comment by jilladams** [comment]
Just noticed you planned to limit this update to dark launches. Is it possible to also reference CMS Feature toggles / flags and their use in content build, in this update? (#5 in the ticket body)

**Comment by jknipes** [comment]
Hi @jilladams, Update: @edmund-dunn is reviewing the entire document (#1-6), so I believe the CMS Feature toggles/flags section is included. 
[Most recent message in thread](https://dsva.slack.com/archives/C03LFSPGV16/p1728330011260259?thread_ts=1726518589.434909&cid=C03LFSPGV16)

**Comment by jilladams** [comment]
Updated step #4 in the Drupal staging-only launch notes above with recent things we learned about more specific mechanisms that must run for deploy to be successful.
